In [1]:
import pandas as pd
pd.options.display.max_columns=50

### Read data

In [2]:
df_income_learn = pd.read_csv('census_income_learn.csv', header=None)
df_income_test = pd.read_csv('census_income_test.csv', header=None)

In [3]:
print(f"Train set size: {len(df_income_learn)}")
print(f"Test set size: {len(df_income_test)}")

Train set size: 199523
Test set size: 99762


In [4]:
# Set column names
df_columns = ['age',
'class_of_worker',
'detailed_industry_recode',
'detailed_occupation_recode',
'education',
'wage_per_hour',
'enroll_in_edu_inst_last_wk',
'marital_stat',
'major_industry_code',
'major_occupation_code',
'race',
'hispanic_origin',
'sex',
'member_of_a_labor_union',
'reason_for_unemployment',
'full_or_part_time_employment_stat',
'capital_gains',
'capital_losses',
'dividends_from_stocks',
'tax_filer_stat',
'region_of_previous_residence',
'state_of_previous_residence',
'detailed_household_and_family_stat',
'detailed_household_summary_in_household',
'instance_weight',
'migration_code-change_in_msa',
'migration_code-change_in_reg',
'migration_code-move_within_reg',
'live_in_this_house_1_year_ago',
'migration_prev_res_in_sunbelt',
'num_persons_worked_for_employer',
'family_members_under_18',
'country_of_birth_father',
'country_of_birth_mother',
'country_of_birth_self',
'citizenship',
'own_business_or_self_employed',
'fill_inc_questionnaire_for_veterans_admin',
'veterans_benefits',
'weeks_worked_in_year',
'year',
'income']

In [5]:
df_income_learn.columns = df_columns
df_income_test.columns = df_columns

In [6]:
df_income_learn.describe()

age  detailed_industry_recode  detailed_occupation_recode  \
count  199523.000000             199523.000000               199523.000000   
mean       34.494199                 15.352320                   11.306556   
std        22.310895                 18.067129                   14.454204   
min         0.000000                  0.000000                    0.000000   
25%        15.000000                  0.000000                    0.000000   
50%        33.000000                  0.000000                    0.000000   
75%        50.000000                 33.000000                   26.000000   
max        90.000000                 51.000000                   46.000000   

       wage_per_hour  capital_gains  capital_losses  dividends_from_stocks  \
count  199523.000000   199523.00000   199523.000000          199523.000000   
mean       55.426908      434.71899       37.313788             197.529533   
std       274.896454     4697.53128      271.896428            1984.163658   
min         0.000000        0.00000        0.000000               0.000000   
25%         0.000000        0.00000        0.000000               0.000000   
50%         0.000000        0.00000        0.000000               0.000000   
75%         0.000000        0.00000        0.000000               0.000000   
max      9999.000000    99999.00000     4608.000000           99999.000000   

       instance_weight  num_persons_worked_for_employer  \
count    199523.000000                    199523.000000   
mean       1740.380269                         1.956180   
std         993.768156                         2.365126   
min          37.870000                         0.000000   
25%        1061.615000                         0.000000   
50%        1618.310000                         1.000000   
75%        2188.610000                         4.000000   
max       18656.300000                         6.000000   

       own_business_or_self_employed  veterans_benefits  weeks_worked_in_year  \
count                  199523.000000      199523.000000         199523.000000   
mean                        0.175438           1.514833             23.174897   
std                         0.553694           0.851473             24.411488   
min                         0.000000           0.000000              0.000000   
25%                         0.000000           2.000000              0.000000   
50%                         0.000000           2.000000              8.000000   
75%                         0.000000           2.000000             52.000000   
max                         2.000000           2.000000             52.000000   

                year  
count  199523.000000  
mean       94.499672  
std         0.500001  
min        94.000000  
25%        94.000000  
50%        94.000000  
75%        95.000000  
max        95.000000

In [7]:
df_income_learn.head()

age                  class_of_worker  detailed_industry_recode  \
0   73                  Not in universe                         0   
1   58   Self-employed-not incorporated                         4   
2   18                  Not in universe                         0   
3    9                  Not in universe                         0   
4   10                  Not in universe                         0   

   detailed_occupation_recode                    education  wage_per_hour  \
0                           0         High school graduate              0   
1                          34   Some college but no degree              0   
2                           0                   10th grade              0   
3                           0                     Children              0   
4                           0                     Children              0   

  enroll_in_edu_inst_last_wk    marital_stat           major_industry_code  \
0            Not in universe         Widowed   Not in universe or children   
1            Not in universe        Divorced                  Construction   
2                High school   Never married   Not in universe or children   
3            Not in universe   Never married   Not in universe or children   
4            Not in universe   Never married   Not in universe or children   

                  major_occupation_code                        race  \
0                       Not in universe                       White   
1   Precision production craft & repair                       White   
2                       Not in universe   Asian or Pacific Islander   
3                       Not in universe                       White   
4                       Not in universe                       White   

  hispanic_origin      sex member_of_a_labor_union reason_for_unemployment  \
0       All other   Female         Not in universe         Not in universe   
1       All other     Male         Not in universe         Not in universe   
2       All other   Female         Not in universe         Not in universe   
3       All other   Female         Not in universe         Not in universe   
4       All other   Female         Not in universe         Not in universe   

  full_or_part_time_employment_stat  capital_gains  capital_losses  \
0                Not in labor force              0               0   
1          Children or Armed Forces              0               0   
2                Not in labor force              0               0   
3          Children or Armed Forces              0               0   
4          Children or Armed Forces              0               0   

   dividends_from_stocks      tax_filer_stat region_of_previous_residence  \
0                      0            Nonfiler              Not in universe   
1                      0   Head of household                        South   
2                      0            Nonfiler              Not in universe   
3                      0            Nonfiler              Not in universe   
4                      0            Nonfiler              Not in universe   

  state_of_previous_residence         detailed_household_and_family_stat  \
0             Not in universe   Other Rel 18+ ever marr not in subfamily   
1                    Arkansas                                Householder   
2             Not in universe    Child 18+ never marr Not in a subfamily   
3             Not in universe      Child <18 never marr not in subfamily   
4             Not in universe      Child <18 never marr not in subfamily   

  detailed_household_summary_in_household  instance_weight  \
0           Other relative of householder          1700.09   
1                             Householder          1053.55   
2                       Child 18 or older           991.95   
3            Child under 18 never married          1758.14   
4            Child under 18 never married          1069.16   

  migration_code-change_in_msa migration_code-

In [8]:
df_income_test.describe()

age  detailed_industry_recode  detailed_occupation_recode  \
count  99762.000000              99762.000000                99762.000000   
mean      34.628596                 15.332812                   11.331118   
std       22.333449                 18.026709                   14.459589   
min        0.000000                  0.000000                    0.000000   
25%       16.000000                  0.000000                    0.000000   
50%       33.000000                  0.000000                    0.000000   
75%       50.000000                 33.000000                   26.000000   
max       90.000000                 51.000000                   46.000000   

       wage_per_hour  capital_gains  capital_losses  dividends_from_stocks  \
count   99762.000000   99762.000000    99762.000000           99762.000000   
mean       54.461268     425.788577       35.919458             192.494727   
std       270.202643    4616.795578      265.525252            1841.728492   
min         0.000000       0.000000        0.000000               0.000000   
25%         0.000000       0.000000        0.000000               0.000000   
50%         0.000000       0.000000        0.000000               0.000000   
75%         0.000000       0.000000        0.000000               0.000000   
max      9900.000000   99999.000000     4608.000000           99999.000000   

       instance_weight  num_persons_worked_for_employer  \
count     99762.000000                     99762.000000   
mean       1739.542838                         1.956156   
std         994.900902                         2.364331   
min          43.260000                         0.000000   
25%        1058.050000                         0.000000   
50%        1615.395000                         1.000000   
75%        2187.277500                         4.000000   
max       16258.200000                         6.000000   

       own_business_or_self_employed  veterans_benefits  weeks_worked_in_year  \
count                   99762.000000       99762.000000          99762.000000   
mean                        0.178505           1.519827             23.185331   
std                         0.558316           0.848229             24.389184   
min                         0.000000           0.000000              0.000000   
25%                         0.000000           2.000000              0.000000   
50%                         0.000000           2.000000              8.000000   
75%                         0.000000           2.000000             52.000000   
max                         2.000000           2.000000             52.000000   

               year  
count  99762.000000  
mean      94.500652  
std        0.500002  
min       94.000000  
25%       94.000000  
50%       95.000000  
75%       95.000000  
max       95.000000

In [9]:
df_income_test.head()

age                  class_of_worker  detailed_industry_recode  \
0   38                          Private                         6   
1   44   Self-employed-not incorporated                        37   
2    2                  Not in universe                         0   
3   35                          Private                        29   
4   49                          Private                         4   

   detailed_occupation_recode                             education  \
0                          36              1st 2nd 3rd or 4th grade   
1                          12   Associates degree-occup /vocational   
2                           0                              Children   
3                           3                  High school graduate   
4                          34                  High school graduate   

   wage_per_hour enroll_in_edu_inst_last_wk                      marital_stat  \
0              0            Not in universe   Married-civilian spouse present   
1              0            Not in universe   Married-civilian spouse present   
2              0            Not in universe                     Never married   
3              0            Not in universe                          Divorced   
4              0            Not in universe                          Divorced   

             major_industry_code                   major_occupation_code  \
0    Manufacturing-durable goods   Machine operators assmblrs & inspctrs   
1   Business and repair services                  Professional specialty   
2    Not in universe or children                         Not in universe   
3                 Transportation          Executive admin and managerial   
4                   Construction     Precision production craft & repair   

     race      hispanic_origin      sex member_of_a_labor_union  \
0   White   Mexican (Mexicano)   Female         Not in universe   
1   White            All other   Female         Not in universe   
2   White     Mexican-American     Male         Not in universe   
3   White            All other   Female         Not in universe   
4   White            All other     Male         Not in universe   

  reason_for_unemployment full_or_part_time_employment_stat  capital_gains  \
0         Not in universe               Full-time schedules              0   
1         Not in universe    PT for econ reasons usually PT              0   
2         Not in universe          Children or Armed Forces              0   
3         Not in universe          Children or Armed Forces              0   
4         Not in universe               Full-time schedules              0   

   capital_losses  dividends_from_stocks                 tax_filer_stat  \
0               0                      0   Joint one under 65 & one 65+   
1               0                   2500            Joint both under 65   
2               0                      0                       Nonfiler   
3               0                      0              Head of household   
4               0                      0                         Single   

  region_of_previous_residence state_of_previous_residence  \
0              Not in universe             Not in universe   
1              Not in universe             Not in universe   
2              Not in universe             Not in universe   
3              Not in universe             Not in universe   
4              Not in universe             Not in universe   

       detailed_household_and_family_stat  \
0                   Spouse of householder   
1                   Spouse of householder   
2   Child <18 never marr not in subfamily   
3                             Householder   
4                    Secondary individual   

  detailed_household_summary_in_household  instance_weight  \
0                   Spouse of householder          1032.38   
1                   Spouse of householder          1462.33   
2            Child under 18 never married          1601.75   
3 

#### Drop unnecessary data
We do not need the instance_weight column, duplicated rows, or the detailed_industry_recode and detailed_occupation_recode columns

In [10]:
# Drop instance weight column
df_income_learn.drop(['instance_weight'], axis=1, inplace=True)
df_income_test.drop(['instance_weight'], axis=1, inplace=True)
print(len(df_income_learn.columns))

41


In [11]:
# Find duplicate rows for train and test
duplicated_learn = df_income_learn[df_income_learn.duplicated()]
print(len(duplicated_learn))

46627


In [12]:
duplicated_test = df_income_test[df_income_test.duplicated()]
print(len(duplicated_test))

20898


In [13]:
# Drop duplicate rows
df_income_learn.drop_duplicates(keep=False,inplace=True)
print(f"Train set size after dropping duplicates: {len(df_income_learn)}")

Train set size after dropping duplicates: 145776


In [14]:
df_income_test.drop_duplicates(keep=False,inplace=True)
print(f"Test set size after dropping duplicates: {len(df_income_test)}")

Test set size after dropping duplicates: 75428


In [15]:
# Look at data after dropping duplicates
df_income_learn.describe()

age  detailed_industry_recode  detailed_occupation_recode  \
count  145776.000000             145776.000000               145776.000000   
mean       40.049946                 20.829862                   15.312788   
std        19.047693                 18.133937                   14.881161   
min         0.000000                  0.000000                    0.000000   
25%        26.000000                  0.000000                    0.000000   
50%        38.000000                 27.000000                   12.000000   
75%        52.000000                 37.000000                   29.000000   
max        90.000000                 51.000000                   46.000000   

       wage_per_hour  capital_gains  capital_losses  dividends_from_stocks  \
count  145776.000000  145776.000000   145776.000000          145776.000000   
mean       75.850915     594.886230       51.011463             268.787709   
std       319.180793    5487.018219      316.587798            2313.427671   
min         0.000000       0.000000        0.000000               0.000000   
25%         0.000000       0.000000        0.000000               0.000000   
50%         0.000000       0.000000        0.000000               0.000000   
75%         0.000000       0.000000        0.000000               0.000000   
max      9999.000000   99999.000000     4608.000000           99999.000000   

       num_persons_worked_for_employer  own_business_or_self_employed  \
count                    145776.000000                  145776.000000   
mean                          2.648776                       0.239381   
std                           2.395149                       0.634786   
min                           0.000000                       0.000000   
25%                           0.000000                       0.000000   
50%                           2.000000                       0.000000   
75%                           5.000000                       0.000000   
max                           6.000000                       2.000000   

       veterans_benefits  weeks_worked_in_year           year  
count      145776.000000         145776.000000  145776.000000  
mean            1.853165             31.365046      94.489223  
std             0.508680             23.437004       0.499886  
min             0.000000              0.000000      94.000000  
25%             2.000000              0.000000      94.000000  
50%             2.000000             49.000000      94.000000  
75%             2.000000             52.000000      95.000000  
max             2.000000             52.000000      95.000000

In [16]:
df_income_test.describe()

age  detailed_industry_recode  detailed_occupation_recode  \
count  75428.000000              75428.000000                75428.000000   
mean      39.902927                 20.166715                   14.892083   
std       19.453138                 18.163667                   14.892245   
min        0.000000                  0.000000                    0.000000   
25%       26.000000                  0.000000                    0.000000   
50%       38.000000                 24.000000                   12.000000   
75%       53.000000                 37.000000                   29.000000   
max       90.000000                 51.000000                   46.000000   

       wage_per_hour  capital_gains  capital_losses  dividends_from_stocks  \
count   75428.000000   75428.000000    75428.000000           75428.000000   
mean       72.031142     563.113777       47.507517             254.493809   
std       308.703640    5302.259899      304.464745            2114.344317   
min         0.000000       0.000000        0.000000               0.000000   
25%         0.000000       0.000000        0.000000               0.000000   
50%         0.000000       0.000000        0.000000               0.000000   
75%         0.000000       0.000000        0.000000               0.000000   
max      9900.000000   99999.000000     4608.000000           99999.000000   

       num_persons_worked_for_employer  own_business_or_self_employed  \
count                     75428.000000                   75428.000000   
mean                          2.573686                       0.235615   
std                           2.401099                       0.630838   
min                           0.000000                       0.000000   
25%                           0.000000                       0.000000   
50%                           2.000000                       0.000000   
75%                           5.000000                       0.000000   
max                           6.000000                       2.000000   

       veterans_benefits  weeks_worked_in_year          year  
count       75428.000000          75428.000000  75428.000000  
mean            1.835989             30.469454     94.491489  
std             0.536472             23.651942      0.499931  
min             0.000000              0.000000     94.000000  
25%             2.000000              0.000000     94.000000  
50%             2.000000             47.000000     94.000000  
75%             2.000000             52.000000     95.000000  
max             2.000000             52.000000     95.000000

**Dropping duplicates seems to be a critical step in data cleaning. For example, after dropping duplicates, the mean age increased from 34.49 to 40.05. This suggests that leaving duplicates in for analysis would have skewed results toward the younger population.**

In [17]:
# Drop detailed_industry_recode and detailed_occupation_recode columns since they are duplicates of major_industry_code and major_occupation_code
df_income_learn.drop(['detailed_industry_recode'], axis=1, inplace=True)
df_income_learn.drop(['detailed_occupation_recode'], axis=1, inplace=True)

df_income_test.drop(['detailed_industry_recode'], axis=1, inplace=True)
df_income_test.drop(['detailed_occupation_recode'], axis=1, inplace=True)

### Next check for null values

In [18]:
df_income_learn.isnull().sum()

age                                          0
class_of_worker                              0
education                                    0
wage_per_hour                                0
enroll_in_edu_inst_last_wk                   0
marital_stat                                 0
major_industry_code                          0
major_occupation_code                        0
race                                         0
hispanic_origin                              0
sex                                          0
member_of_a_labor_union                      0
reason_for_unemployment                      0
full_or_part_time_employment_stat            0
capital_gains                                0
capital_losses                               0
dividends_from_stocks                        0
tax_filer_stat                               0
region_of_previous_residence                 0
state_of_previous_residence                  0
detailed_household_and_family_stat           0
detailed_hous

In [19]:
df_income_test.isnull().sum()

age                                          0
class_of_worker                              0
education                                    0
wage_per_hour                                0
enroll_in_edu_inst_last_wk                   0
marital_stat                                 0
major_industry_code                          0
major_occupation_code                        0
race                                         0
hispanic_origin                              0
sex                                          0
member_of_a_labor_union                      0
reason_for_unemployment                      0
full_or_part_time_employment_stat            0
capital_gains                                0
capital_losses                               0
dividends_from_stocks                        0
tax_filer_stat                               0
region_of_previous_residence                 0
state_of_previous_residence                  0
detailed_household_and_family_stat           0
detailed_hous

**There are no null values**

### We can bin the education column to reduce the number of categories

In [20]:
df_income_learn['education'].value_counts()

 High school graduate                      42206
 Some college but no degree                25740
 Bachelors degree(BA AB BS)                19137
 Children                                   9742
 Masters degree(MA MS MEng MEd MSW MBA)     6396
 10th grade                                 6114
 11th grade                                 5950
 7th and 8th grade                          5766
 Associates degree-occup /vocational        5173
 9th grade                                  4649
 Associates degree-academic program         4295
 5th or 6th grade                           3041
 12th grade no diploma                      2016
 Prof school degree (MD DDS DVM LLB JD)     1789
 1st 2nd 3rd or 4th grade                   1716
 Doctorate degree(PhD EdD)                  1261
 Less than 1st grade                         785
Name: education, dtype: int64

**We can bin education to up to 6th grade and high school without graduating**

In [21]:
up_to_6th_grade = [' 5th or 6th grade', ' 1st 2nd 3rd or 4th grade', ' Less than 1st grade']
high_school_no_graduate = [' 9th grade', ' 10th grade', ' 11th grade', ' 12th grade no diploma']

for column in up_to_6th_grade:
    df_income_learn['education'] = df_income_learn['education'].replace(column, 'up_to_6th_grade')

for column in high_school_no_graduate:
    df_income_learn['education'] = df_income_learn['education'].replace(column, 'high_school_no_graduate')


In [22]:
df_income_learn['education'].value_counts()

 High school graduate                      42206
 Some college but no degree                25740
 Bachelors degree(BA AB BS)                19137
high_school_no_graduate                    18729
 Children                                   9742
 Masters degree(MA MS MEng MEd MSW MBA)     6396
 7th and 8th grade                          5766
up_to_6th_grade                             5542
 Associates degree-occup /vocational        5173
 Associates degree-academic program         4295
 Prof school degree (MD DDS DVM LLB JD)     1789
 Doctorate degree(PhD EdD)                  1261
Name: education, dtype: int64

In [23]:
for column in up_to_6th_grade:
    df_income_test['education'] = df_income_test['education'].replace(column, 'up_to_6th_grade')
for column in high_school_no_graduate:
    df_income_test['education'] = df_income_test['education'].replace(column, 'high_school_no_graduate')

In [24]:
df_income_test['education'].value_counts()

 High school graduate                      21535
 Some college but no degree                13135
high_school_no_graduate                     9964
 Bachelors degree(BA AB BS)                 9618
 Children                                   5687
 Masters degree(MA MS MEng MEd MSW MBA)     3254
 7th and 8th grade                          3157
up_to_6th_grade                             2910
 Associates degree-occup /vocational        2631
 Associates degree-academic program         2043
 Prof school degree (MD DDS DVM LLB JD)      874
 Doctorate degree(PhD EdD)                   620
Name: education, dtype: int64

### Finally, replace the income label with binary integer values

In [25]:
# Replace income values with binary integer values
df_income_learn['income'] = df_income_learn['income'].replace(" - 50000.", 0) 
df_income_learn['income'] = df_income_learn['income'].replace(" 50000+.", 1)
df_income_learn.income.value_counts()

0    133424
1     12352
Name: income, dtype: int64

In [26]:
df_income_test['income'] = df_income_test['income'].replace(" - 50000.", 0) 
df_income_test['income'] = df_income_test['income'].replace(" 50000+.", 1)
df_income_test.income.value_counts()

0    69242
1     6186
Name: income, dtype: int64

### Pickle dataframes for further analysis

In [27]:
df_income_learn.to_pickle('data/df_income_learn.pkl')

In [28]:
df_income_test.to_pickle('data/df_income_test.pkl')